File: typescript/test/useCase/OrderApprovalUseCase.test.ts
import Order from '../../src/domain/Order';
import { OrderStatus } from '../../src/domain/OrderStatus';
import ApprovedOrderCannotBeRejectedException from '../../src/useCase/ApprovedOrderCannotBeRejectedException';
import OrderApprovalRequest from '../../src/useCase/OrderApprovalRequest';
import OrderApprovalUseCase from '../../src/useCase/OrderApprovalUseCase';
import RejectedOrderCannotBeApprovedException from '../../src/useCase/RejectedOrderCannotBeApprovedException';
import ShippedOrdersCannotBeChangedException from '../../src/useCase/ShippedOrdersCannotBeChangedException';
import TestOrderRepository from '../doubles/TestOrderRepository';

describe('OrderApprovalUseCase', () => {
  let orderRepository: TestOrderRepository;
  let useCase: OrderApprovalUseCase;

  beforeEach( () => {
    orderRepository = new TestOrderRepository();
    useCase = new OrderApprovalUseCase(orderRepository);
  });
  it('approvedExistingOrder', () => {
    let initialOrder: Order = new Order();
    initialOrder.setStatus(OrderStatus.CREATED);
    initialOrder.setId(1);
    orderRepository.addOrder(initialOrder);

    let request: OrderApprovalRequest = new OrderApprovalRequest();
    request.setOrderId(1);
    request.setApproved(true);

    useCase.run(request);

    const savedOrder: Order = orderRepository.getSavedOrder();
    expect(savedOrder.getStatus()).toBe(OrderStatus.APPROVED);
  });

  it('rejectedExistingOrder', () => {
    let initialOrder: Order = new Order();
    initialOrder.setStatus(OrderStatus.CREATED);
    initialOrder.setId(1);
    orderRepository.addOrder(initialOrder);

    let request: OrderApprovalRequest = new OrderApprovalRequest();
    request.setOrderId(1);
    request.setApproved(false);

    useCase.run(request);

    const savedOrder: Order = orderRepository.getSavedOrder();
    expect(savedOrder.getStatus()).toBe(OrderStatus.REJECTED);
  });

  it('cannotApproveRejectedOrder', () => {
    const initialOrder: Order = new Order();
    initialOrder.setStatus(OrderStatus.REJECTED);
    initialOrder.setId(1);
    orderRepository.addOrder(initialOrder);

    const request: OrderApprovalRequest = new OrderApprovalRequest();
    request.setOrderId(1);
    request.setApproved(true);

    expect(() => useCase.run(request)).toThrow(RejectedOrderCannotBeApprovedException);
    expect(orderRepository.getSavedOrder()).toBe(null);
  });

  it('cannotRejectApprovedOrder', () => {
    const initialOrder: Order = new Order();
    initialOrder.setStatus(OrderStatus.APPROVED);
    initialOrder.setId(1);
    orderRepository.addOrder(initialOrder);

    const request: OrderApprovalRequest = new OrderApprovalRequest();
    request.setOrderId(1);
    request.setApproved(false);

    expect(() =>  useCase.run(request)).toThrow(ApprovedOrderCannotBeRejectedException);
    expect(orderRepository.getSavedOrder()).toBe(null);
  });

  it('shippedOrdersCannotBeApproved', () => {
    const initialOrder: Order = new Order();
    initialOrder.setStatus(OrderStatus.SHIPPED);
    initialOrder.setId(1);
    orderRepository.addOrder(initialOrder);

    const request: OrderApprovalRequest = new OrderApprovalRequest();
    request.setOrderId(1);
    request.setApproved(true);

    expect(() => useCase.run(request)).toThrow(ShippedOrdersCannotBeChangedException);
    expect(orderRepository.getSavedOrder()).toBe(null);
  });

  it('shippedOrdersCannotBeRejected', () => {
    let initialOrder: Order = new Order();
    initialOrder.setStatus(OrderStatus.SHIPPED);
    initialOrder.setId(1);
    orderRepository.addOrder(initialOrder);

    let request: OrderApprovalRequest = new OrderApprovalRequest();
    request.setOrderId(1);
    request.setApproved(false);

    expect(() => useCase.run(request)).toThrow(ShippedOrdersCannotBeChangedException);
    expect(orderRepository.getSavedOrder()).toBe(null);
  });
});

File: scala/src/test/scala/usecase/OrderApprovalUseCaseTest.scala
package usecase

import doubles.TestOrderRepository
import ordershipping.domain.{Order, OrderStatus}
import ordershipping.usecase.{
  ApprovedOrderCannotBeRejectedException,
  OrderApprovalRequest,
  OrderApprovalUseCase,
  RejectedOrderCannotBeApprovedException,
  ShippedOrdersCannotBeChangedException
}
import org.scalatest.BeforeAndAfterEach
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers

class OrderApprovalUseCaseTest
    extends AnyFlatSpec
    with Matchers
    with BeforeAndAfterEach {
  private var orderRepository: TestOrderRepository = _
  private var useCase: OrderApprovalUseCase = _

  override def beforeEach(): Unit = {
    orderRepository = new TestOrderRepository
    useCase = new OrderApprovalUseCase(orderRepository)
  }

  "order approval use case" should "approved existing order" in {
    val initialOrder = new Order(status = OrderStatus.Created, id = 1)
    orderRepository.addOrder(initialOrder)
    val request = OrderApprovalRequest(orderId = 1, approved = true)

    useCase.run(request)

    val savedOrder = orderRepository.savedOrder()
    savedOrder.status shouldBe OrderStatus.Approved
  }

  "order approval use case" should "rejected existing order" in {
    val initialOrder = new Order(status = OrderStatus.Created, id = 1)
    orderRepository.addOrder(initialOrder)
    val request = OrderApprovalRequest(orderId = 1, approved = false)

    useCase.run(request)

    val savedOrder = orderRepository.savedOrder()
    savedOrder.status shouldBe OrderStatus.Rejected
  }

  "order approval use case" should "can not approve rejected order" in {
    val initialOrder = new Order(status = OrderStatus.Rejected, id = 1)
    orderRepository.addOrder(initialOrder)
    val request = OrderApprovalRequest(orderId = 1, approved = true)

    assertThrows[RejectedOrderCannotBeApprovedException] {
      useCase.run(request)
    }
    orderRepository.savedOrder() shouldBe null
  }

  "order approval use case" should "can not reject approved order" in {
    val initialOrder = new Order(status = OrderStatus.Approved, id = 1)
    orderRepository.addOrder(initialOrder)
    val request = OrderApprovalRequest(orderId = 1, approved = false)

    assertThrows[ApprovedOrderCannotBeRejectedException] {
      useCase.run(request)
    }
    orderRepository.savedOrder() shouldBe null
  }

  "order approval use case" should "can not reject shipped order" in {
    val initialOrder = new Order(status = OrderStatus.Shipped, id = 1)
    orderRepository.addOrder(initialOrder)
    val request = OrderApprovalRequest(orderId = 1, approved = false)

    assertThrows[ShippedOrdersCannotBeChangedException] {
      useCase.run(request)
    }
    orderRepository.savedOrder() shouldBe null
  }
}
